rl_chain/rl_chain_base.py

```diff
@@ -282,38 +282,54 @@ def output_keys(self) -> List[str]:
         return [self.output_key]
 
     @abstractmethod
-    def call_before_predict(self, inputs: Dict[str, Any]) -> Event:
+    def _call_before_predict(self, inputs: Dict[str, Any]) -> Event:
         pass
 
     @abstractmethod
-    def call_after_predict_before_llm(
+    def _call_after_predict_before_llm(
         self, inputs: Dict[str, Any], event: Event, vwpreds: Any
     ) -> Tuple[Dict[str, Any], Event]:
         pass
 
     @abstractmethod
-    def call_after_llm_before_scoring(
+    def _call_after_llm_before_scoring(
         self, llm_response: str, event: Event
     ) -> Tuple[Dict[str, Any], Event]:
         pass
 
     @abstractmethod
-    def call_after_scoring_before_learning(
-        self, llm_response: str, event: Event, response_quality: Optional[float]
+    def _call_after_scoring_before_learning(
+        self, event: Event, response_quality: Optional[float]
     ) -> Event:
         pass
 
+    def learn_delayed_reward(
+        self, reward: float, event: Event, force_reward=False
+    ) -> None:
+        """
+        Learn will be called with the reward specified and the actions/embeddings/etc stored in event
+
+        Will raise an error if response_validator is set, and force_reward=True was not provided during the method call
+        """
+        if self.response_validator and not force_reward:
+            raise RuntimeError(
+                "The response validator is set, and force_reward was not set to True. Please set force_reward=True to use this function."
+            )
+        self._call_after_scoring_before_learning(event=event, response_quality=reward)
+        self.policy.learn(event=event)
+        self.policy.log(event=event)
```

Comment:
hm not sure about policy log here, if logging is enabled will log once without cost and once with cost now (once in main call), maybe this is ok